scripts/createLanguageServiceBuild.ts

```diff
@@ -0,0 +1,46 @@
+/// <reference types="node"/>
+import { join } from "path";
+import { readFileSync, unlinkSync } from "fs";
+import { tmpdir } from "os";
+import { execSync, ExecSyncOptions } from "child_process";
+import chalk from "chalk";
+
+interface PackageJson {
+    name: string;
+    version: string
+}
+
+const exec = (cmd: string, opts?: ExecSyncOptions) => {
+    console.log(chalk.gray(`> ${cmd} ${opts ? JSON.stringify(opts) : ""}`));
+    execSync(cmd, opts);
+};
+
+const step = (msg: string) => {
+    console.log("\n\n" + chalk.bold("- ") + msg);
+};
+
+function main(): void {
+    console.log(chalk.bold("## Creating the language services build of TypeScript"));
+    process.stdout.write(chalk.grey("> node /scripts/createLanguageServiceBuild.ts"));
+
+    // Create a tarball of the current version
+    step("Packing the current TypeScript via npm.");
+    exec("npm pack");
+
+    const packageJsonValue: PackageJson = JSON.parse(readFileSync("package.json", "utf8"));
+    const tarballFileName = `${packageJsonValue.name}-${packageJsonValue.version}.tgz`;
+
+    const unzipDir = tmpdir();
+    step(`Extracting the built version into a temporary folder. ${unzipDir}/package`);
+    exec(`tar -xvzf ${tarballFileName} -C ${unzipDir}`);
+    unlinkSync(tarballFileName);
+
+    step(`Updating the build metadata`);
+    const packagePath = join(unzipDir, "package");
+    exec(`node scripts/configureLanguageServiceBuild.js ${join(packagePath, "package.json")}`);
+
+    step(`Deploying the language service`);
+    exec("npm publish --access public", { cwd: packagePath });

```

Comment:
Do we have any other scripts that auto-publish? I don’t know what our publish process is like, but I assume it runs in CI. I would lean toward adding automation there rather than scripted into `postpublish`, as principle of least surprise.

Reply:
IOW, the code seems fine (but I’ve never looked at our packing process before either, so I’m not the best one to review), but I’m skeptical of the trigger

Reply:
Nope, it doesn't happen on CI (I am happy to move it to a new step in there) and/or look at moving to deploy _on_ CI ( e.g https://github.com/microsoft/types-publisher/pull/659 ) and all we do is ship a tag to indicate we want a release

Reply:
Oh really? I figured we publish from here: https://dev.azure.com/typescript/TypeScript/_release?definitionId=1&view=mine&_a=releases

Reply:
We do! It definitely happens on CI! Anyone not publishing on CI will be [redacted]!

Reply:
Specifically: For all `release-x` branches, the node 12 build always produces a tarball based on the branch. Any one of those tarballs and be promoted (from the build UI) and published as a full release, which then triggers `npm publish`es and `github` releases. So the workflow is to prep the release branch, bump it's version, do an LKG, then fire off that generated tarball as the release.

At no point is an `npm publish` ever run on a folder in our repo ;)

Reply:
Specifically, we publish a prepared tarball _directly_, using `npm publish $(System.ArtifactsDirectory)/typescript.tgz --tag $(PUBLISH_TAG)`

Reply:
So a `postpublish` script is actually wholly nonfunctional, since we don't even publish in a clone of our repo 😄

Reply:
Hah! OK cool, perfect!

I'll that and add the extra step on CI 👯‍♂

Reply:
Cool - updated the deployment docs, and removed the auto-publish. Will give this a few hours for last feedback and merge today. Then once that's in I'll add the step to the pipeline.

Reply:
This script is _not executable_ in the release pipeline, as far as I'm aware. Releases are not executed within the context of the repo. (There's good reason for this, too, it means at no point does potentially malicious code get a chance to execute in a context where we have our publishing key set!) The code in the repo _does not exist_ in the release pipeline. To publish a second package, you would need to publish a _second_ pipeline artifact with the secondary tarball (which is just done with a few commands in the build pipeline, but I imagine you'll need to add a call to the `configure` script in addition to replicating those), and publish in a similar way as we do the first (with the `npm publish` release pipeline task).